[FSDP2] Move to public `torch.distributed.fsdp`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141868

**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Changes for Reland**
- Preserved the public objects from `torch/distributed/_composable/fsdp/fully_shard.py` so that the import path still works internally
- Added a unit test that we can do `from torch.distributed._composable.fsdp.fully_shard import FSDPModule`

cc @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

Differential Revision: [D66890387](https://our.internmc.facebook.com/intern/diff/D66890387)

Commit:
[FSDP2] Move to public `torch.distributed.fsdp`

[ghstack-poisoned]

awgu commented on test/distributed/_composable/test_composability/test_pp_composability.py:
```diff
@@ -6,17 +6,14 @@
 from torch.distributed.checkpoint.state_dict_loader import _load_state_dict
 from torch.distributed.checkpoint.stateful import Stateful
 from torch.distributed.device_mesh import init_device_mesh
+from torch.distributed.fsdp.fully_shard import fully_shard, MixedPrecisionPolicy
```
```suggestion
from torch.distributed.fsdp import fully_shard, MixedPrecisionPolicy
```

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py:
```diff
@@ -8,10 +8,8 @@
-    _get_gradient_divide_factors,
-)
+from torch.distributed.fsdp import fully_shard, MixedPrecisionPolicy
+from torch.distributed.fsdp._fsdp_collectives import _get_gradient_divide_factors
```
```suggestion
from torch.distributed.fsdp.fully_shard._fsdp_collectives import _get_gradient_divide_factors
```

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"




cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


will need several CI iterations to get this right -- please do not review yet

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release.

**Follow-Ups**
- Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_shard/fully_shard.py:
```diff
@@ -50,41 +58,50 @@ def fully_shard(
         mesh (Optional[DeviceMesh]): This data parallel mesh defines the
             sharding and device. If 1D, then parameters are fully sharded
-            across the 1D mesh (FSDP). If 2D, then parameters are sharded
-            across the 0th dim and replicated across the 1st dim (HSDP). The
```
This was wrong before.

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_init.py:
```diff
@@ -1156,5 +1156,26 @@ def shard_placement_fn(param: nn.Parameter) -> Optional[Shard]:
 
+# TODO: Remove this test class once we remove the old import path:
+# torch/distributed/_composable/fsdp
+class TestFullyShardOldImport(FSDPTestMultiThread):
```
Added this to be sure that old import works without error

awgu commented:
@yf225 I had to make a few path changes across the unit test, Dynamo, and Inductor. They should be safe changes validated by the unit tests, but feel free to take a look at those parts!

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

svekars commented on docs/source/distributed.fsdp.fully_shard.rst:
```diff
@@ -0,0 +1,40 @@
+
```
```suggestion
```

svekars committed:
Update docs/source/distributed.fsdp.fully_shard.rst

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- Add some explanation in the docs about FSDP1 vs. FSDP2
- Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

fegin commented on docs/source/distributed.fsdp.fully_shard.rst:
```diff
@@ -0,0 +1,54 @@
+  sharding representation compared to FSDP1's flat-parameter sharding, while
+  preserving similar throughput performance. This provides a more intuitive
+  user experience, relaxes constraints around frozen parameters, and allows for
+  communication-free sharded state dicts.
```
Should we also add that `full state dict` is not supported by FSDP2?

awgu replied:
Added a bullet

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- [x] Add some explanation in the docs about FSDP1 vs. FSDP2
- [ ] Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

kwen2501 commented on torch/distributed/_composable/fsdp/__init__.py:
```diff
@@ -1,2 +1,8 @@
+    MixedPrecisionPolicy,
+    OffloadPolicy,
+    register_fsdp_forward_method,
+)
```
Perhaps added a redirect warning here saying `_composable.fsdp` is now moved to `torch.distributed.fsdp`?

awgu replied:
Good point! I was thinking we can land a follow-up diff first to migrate the internal callsites and then add a warning for OSS. I wanted to avoid showing the warning to some existing internal customers, which might be disconcerting.

kwen2501 reviewed (approved):
I reviewed the changes in `torch/distributed`. LGTM. Congrats!

wconstab commented on docs/source/distributed.fsdp.fully_shard.rst:
```diff
@@ -0,0 +1,66 @@
+targeting performant eager-mode while using per-parameter sharding for improved
+usability. Compared to PyTorch FSDP1 (``FullyShardedDataParallel``):
+
+- FSDP2 uses ``DTensor``-based dim-0 per-parameter sharding for a simpler
```
nit: should we define what 'per-parameter sharding' actually is?  (mention how copy-in is used, etc)  - or redirect to another wiki where this is explained more in depth?

wconstab commented on docs/source/distributed.fsdp.fully_shard.rst:
```diff
@@ -0,0 +1,66 @@
+  sharding representation compared to FSDP1's flat-parameter sharding, while
+  preserving similar throughput performance. This provides a more intuitive
+  user experience, relaxes constraints around frozen parameters, and allows for
+  communication-free sharded state dicts.
```
nit: would like to say more about the drawbacks of flat-param sharding, esp. RE state-dicts.  again, maybe OK to link to another wiki or post where this is more in depth.  (but in particular, this is a motivating factor for switching)

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- [x] Add some explanation in the docs about FSDP1 vs. FSDP2
- [ ] Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

svekars commented on docs/source/distributed.fsdp.fully_shard.rst:
```diff
@@ -0,0 +1,74 @@
+torch.distributed.fsdp.fully_shard
+==================================
+
+PyTorch FSDP2 (``fully_shard``)
```
This can be done in a follow up review but we need to outline the difference between FSDP1 and FSDP2 - when would I use either of them or whether FSDP2 from now a preferred way. Possibly, add a note in the torch.distributed.fsdp that if you are starting from scratch, use FSDP2 or something along those lines.

awgu replied:
Are you mainly looking for the explicit recommendation of when to use FSDP1 vs. when to use FSDP2?

There is some comparison of FSDP1 vs. FSDP2 in this PR.

awgu replied:
I added a sentence.

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- [x] Add some explanation in the docs about FSDP1 vs. FSDP2
- [ ] Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- [x] Add some explanation in the docs about FSDP1 vs. FSDP2
- [ ] Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

awgu commented:
@pytorchbot merge

atalman commented:
@pytorchmergebot revert -c ghfirst -m "failing internally"

Commit:
Update on "[FSDP2] Move to public `torch.distributed.fsdp`"


**Overview**
This PR moves `torch/distributed/_composable/fsdp` to `torch/distributed/fsdp/_fully_shard` and makes public APIs available from `torch.distributed.fsdp`, e.g.:
```
from torch.distributed.fsdp import fully_shard
```
This is targeting 2.6 release. I rewrote some of the documentation with (hopefully) improved phrasing.

**Follow-Ups**
- [x] Add some explanation in the docs about FSDP1 vs. FSDP2
- [ ] Move unit tests from `test/distributed/_composable/fsdp` to `test/distributed/fsdp/fully_shard/`


cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov LucasLLC MeetVadakkanchery mhorowitz pradeepfn

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66890387).

awgu commented:
```
test_nestedtensor.py::TestNestedTensorOpInfoCUDA::test_compile_backward_nn_functional_linear_cuda_float32
```
failure is not related

awgu commented:
internal tests look good!

awgu commented:
@pytorchbot merge -i
